### PR TITLE
[IMP] website: ignore anchor but not qs when comparing menu URL

### PR DESCRIPTION
**Pull request by rdeodoo**
- `clean_url` method was wrongly flag as `api.model`
- the `unslug_url` was not called for the URL comparison in the dropdown
  case, meaning that `/shop/prod-1` would not match `/shop/product-1` as
  it should (and as it does for regular non dropdown menu)
- the `active` class was actually never working for the dropdown case,
  as the class was added on the wrong element (`li` instead of `a`)

The code was hard to read (mainly because huge python conditions in XML)
and kinda redundant, going through an util method should be clearer and
help reading the template XML.

- Ignore anchors, those are not sent to the server anyway, no way to
  compare even if we wanted to
- Ensure query string (qs) are the same to be considered equals

On top of that, it also fixes the case when the user inserted an
absolute URL instead of a relative one, it will now match.

task-3096367
opw-3091427

### CODE REVIEW

**Comment by qsm-odoo** [comment]
@rdeodoo Seeing the XML changes, I would say master. But if you think it's ok to merge for 14.0, I don't mind. In stable versions, don't users have alternatives at least for the query string part? Like maybe creating a redirect to fool the system? Or maybe it could be a JS solution? :shrug: Anyways, I'd say the less time you lose on this the better, so I would just target master (let's not over-react to one user feedback on something non critical that was not reporter for years ;))... but as you wish :)

**Comment by rdeodoo** [comment]
Added some python tests and moved to master

File: addons/website/models/website_menu.py [hunk 1]
```diff
@@ -136,6 +137,26 @@ def clean_url(self):
                     url = '/%s' % self.url
         return url
 
+    def is_active(self):
```
Comment (qsm-odoo) [inline addons/website/models/website_menu.py:142]: Should it be `_is_active` to not expose it as a callable public method? Probably the same for `clean_url` ?

File: addons/website/models/website_menu.py [hunk 2]
```diff
@@ -136,6 +137,26 @@ def clean_url(self):
...
+        if not menu_url:
+            return False
+
+        menu_url = unslug_url(menu_url)
+        request_url = unslug_url(request.httprequest.path)
+        if request_url == menu_url:
+            return True
+
+        if menu_url != '/' and any(request_url == unslug_url(child.url) for child in self.child_id if child.url):
+            return True
```
Comment (qsm-odoo) [inline addons/website/models/website_menu.py:177]: Maybe a comment above this one. Not entirely sure what it handles: dropdown-toggles ?

File: addons/website/views/website_templates.xml [hunk 3]
```diff
@@ -5,18 +5,16 @@
     <t t-set="show_dropdown" t-value="(submenu.is_mega_menu and submenu.is_visible) or submenu.child_id.filtered(lambda menu: menu.is_visible)"/>
     <li t-if="submenu.is_visible and not (submenu.child_id or submenu.is_mega_menu)" t-attf-class="#{item_class or ''}">
         <a t-att-href="submenu.clean_url()"
-            t-attf-class="#{link_class or ''} #{'active' if submenu.clean_url() and unslug_url(request.httprequest.path) == unslug_url(submenu.clean_url()) else ''}"
+            t-attf-class="#{link_class or ''} #{'active' if submenu.is_active() else ''}"
```
Comment (qsm-odoo) [inline addons/website/views/website_templates.xml:8]: Maybe take the opportunity to write the condition as `#{submenu.is_active() and 'active'}"` ? Same below (or / else unnecessary).

File: addons/website/models/website_menu.py [hunk 4]
```diff
@@ -147,12 +164,16 @@ def is_active(self):
...
+        request_url = url_parse(unslug_url(request.httprequest.url))
+        if menu_url.path == request_url.path and menu_url.decode_query() == request_url.decode_query():
+            if menu_url.netloc and menu_url.netloc != request_url.netloc:
+                # correct path but not correct domain
+                return False
             return True
 
-        if menu_url != '/' and any(request_url == unslug_url(child.url) for child in self.child_id if child.url):
+        if menu_url != '/' and any(child.is_active() for child in self.child_id):
+            # Child match
```
Comment (qsm-odoo) [inline addons/website/models/website_menu.py:176]: Mmh there is one comment in the end, I guess it is enough :)

File: addons/website/models/website_menu.py [hunk 5]
```diff
@@ -147,12 +164,16 @@ def is_active(self):
...
 
-        menu_url = unslug_url(menu_url)
-        request_url = unslug_url(request.httprequest.path)
-        if request_url == menu_url:
+        menu_url = url_parse(unslug_url(menu_url))
+        request_url = url_parse(unslug_url(request.httprequest.url))
+        if menu_url.path == request_url.path and menu_url.decode_query() == request_url.decode_query():
+            if menu_url.netloc and menu_url.netloc != request_url.netloc:
+                # correct path but not correct domain
+                return False
```
Comment (qsm-odoo) [inline addons/website/models/website_menu.py:172]: - Is it necessary to nest this if in the other one? If yes, maybe a comment because I don't understand why.
- Should we actually return False in that case? If fact, even in the case "not menu_url" above, should not we let it go though up to the children comparison as currently an URL on a menu with children does not make sense? :grimacing:

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/website/models/website_menu.py (hunk 1 above)
Comment (rdeodoo) [reply addons/website/models/website_menu.py]: Sure, changed

File: addons/website/models/website_menu.py (hunk 2 above)
Comment (rdeodoo) [reply addons/website/models/website_menu.py]: There is one in next commit as you mentioned at https://github.com/odoo/odoo/pull/107782/commits/25ea3e7f20edc435fd933d084b3adadb13c03412#r1059395918 (on top of a full docstring)
But I'll move it in the first commit

File: addons/website/views/website_templates.xml (hunk 3 above)
Comment (rdeodoo) [reply addons/website/views/website_templates.xml]: Sure, I kept the minimal diff but I'll change

File: addons/website/models/website_menu.py (hunk 1 above)
Comment (rdeodoo) [reply addons/website/models/website_menu.py]: I guess I should create a migration script for the `clean_url` changed tho, as this is typically called from templates, like in our internal repo

File: addons/website/models/website_menu.py (hunk 5 above)
Comment (rdeodoo) [reply addons/website/models/website_menu.py]: For first point, do you ask if I can merge both if and let this case fallback on the final return of the method?
```py
# [...]
if abc:
    if xyz:
        return False
    return True
# [...]
return False
```
-> 
```py
# [...]
if abc and not xyz:
    return True
# [...]
return False
```
Yes it is possible but I don't like that because it's a specific corner case, I like it encapsulated in its if with a comment in the main if, it makes it easier to read and less prone to a later bug where one would add a condition below before the final return that would make this case go through it too

> maybe a comment because I don't understand why.

There is one: `# correct path but not correct domain`, let me know what should I make more clear, right now it's quite "simple": if the path and the query string match, return true except if the domain don't match (which is the corner case)

For second point, there was indeed a case not handled: we don't compare the menu url **at all** if it has submenus. I fixed that and added quite some tests to cover everything

File: addons/http_routing/models/ir_http.py [hunk 6]
```diff
@@ -104,8 +104,9 @@ def slug(value):
 
 
 def unslug(s):
-    """Extract slug and id from a string.
+    """ Extract slug and id from a string.
         Always return un 2-tuple (str|None, int|None)
+        Note that any query string or anchor will prevent the unslug
```
[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/http_routing/models/ir_http.py:109]: Hm that's right.
Due to the `(?=$|/)` end part of the regex, this is only true when there is no `/` before the qs or anchor (which is probably 99% of the URLs, especially in Odoo where we try to enforce non trailing slash URLs).

Here is the output of the current regex matching:
image

It should probably be changed to  `(?=$|\/|#|\?)` (just adding those alternative `|#|\?`) so it also catch those qs/anchor cases. 
image

I'll add a commit about that.

File: addons/website/models/website_menu.py [hunk 7]
```diff
@@ -138,6 +140,49 @@ def clean_url(self):
                     url = '/%s' % self.url
         return url
 
+    def _is_active(self):
+        """ To be considered active, a menu should either:
+        - have its URL matching the request's URL and have no children
```
[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website/models/website_menu.py:145]: I am not sure to understand, could you explain? Do you mean after `"""` or before `- `?

OnPaste 20230117-150023

File: addons/http_routing/models/ir_http.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/http_routing/models/ir_http.py]: Commit + tests added for that.
Probably something that could land in stable but maybe not worth it (low reward / low risk)

File: addons/website/models/website_menu.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website/models/website_menu.py:145]: It's from vscode.
I see what you mean, I had a look by building the doc locally and:
image

After new line:
image

Anyway, it's adapted, thank you about that, I knew about new line before (as in markdown to not make following stuff part of the list) but I didn't know about next line in rst (I usually blindly follow what I find in models.py :grimacing: )

[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website/models/website_menu.py:146]: It's using markdown, and it's not possible to change that AFAIK. Seems to be a recurring problem.
Useful links:
- https://github.com/microsoft/pylance-release/issues/1848#issuecomment-456096029
- https://github.com/microsoft/pylance-release/issues/1848#issuecomment-376278440
- https://github.com/microsoft/pyright/blob/main/packages/pyright-internal/src/analyzer/docStringConversion.ts

Anyway, I'll pay attention to those list formatting.

File: addons/http_routing/models/ir_http.py [hunk 8]
```diff
@@ -114,7 +114,9 @@ def unslug(s):
 
 
 def unslug_url(s):
-    """ From /blog/my-super-blog-1" to "blog/1" """
+    """ From /blog/my-super-blog-1" to "blog/1"
+        Note that any query string or anchor will prevent the unslug
```
Comment (qsm-odoo) [inline addons/http_routing/models/ir_http.py:118]: Is not that what you fixed in the first commit? I tested:
```
In [4]: unslug_url('hello/world-2?foo=bar')
Out[4]: 'hello/2'
```
It seems to lose the querystring but not prevent the unslug

File: addons/website/models/website_menu.py [hunk 9]
```diff
@@ -139,24 +141,47 @@ def _clean_url(self):
...
+        - or have a children menu URL matching the request's URL
+
+        Matching an URL means, either:
+
+        - be equal, eg ``/contact/on-site`` vs ``/contact/on-site``
+        - be equal after unslug, eg ``/shop/1`` and ``/shop/my-super-product-1``
+
+        Note that saving a menu URL with an anchor or a query string is
+        considered a corner case, and the following heuristic will be applied:
```
Comment (qsm-odoo) [inline addons/website/models/website_menu.py:155]: Nitpicking: "the following applies:" instead of "the following heuristic will be applied:". It kinda was an heuristic before but not anymore IMO :grimacing:

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
Thanks for the test and fix :) A quick comment then I'll r+ :)

File: addons/http_routing/models/ir_http.py (hunk 8 above)
Comment (rdeodoo) [reply addons/http_routing/models/ir_http.py:118]: Indeed, it's what I fixed in the first commit that landed afterward, as I was sad keeping this misbehavior, despite commenting it.
This should have thus been removed as not true anymore, good catch.
